Propose additional JVMs in Java buildpack (#267)
paketo-buildpacks/rfcs: text/java/0013-alternate-jvms-in-java-buildpack.md

## Motivation

There are a few motivations for this change:

1. Using an alternate JVM is tedious.  
2. The preferred JVM for the Liberty application server is eclipse openj9 because there are features in it that the liberty server takes advantage of we would like to exploit in the liberty buildpack.
3. The alternate JVM buildpack runs before the CA certs buildpack and therefore traffic from the alternate JVM vendor buildpack won’t trust any additional CA certs.  
4. Easier adoption of the other JVM buildpacks.

## Detailed Explanation

Currently, the Java buildpack only contributes the BellSoft Liberica JVM.  If we can accommodate multiple JVMs then we can provide a more
well-rounded java buildpack, reach more Java communities, generate more interest in Paketo and Cloud Native buildpacks, and ultimately make it easier for users to build images for Java applications.

## Rationale and Alternatives

Alternatives:

- Do nothing. Users of the Java buildpack will have to manually create a builder or manually specify the buildpack order, which is tedious.
- Create new composite buildpacks other JVMs as the default.

## Implementation

> In addition to the current criteria for each buildpack, the detect method will also validate itself against the value of `BP_JAVA_JVM` and contribute only if it matches. BellSoft Liberica will remain the default JVM.

Author: I just threw a name out there without giving it much thought.  I'm don't have any attachment to any name.    

`BP_JAVA_VENDOR` makes sense.

> In addition to the current criteria for each buildpack, the detect method will also validate itself against the value of `BP_JAVA_JVM` and contribute only if it matches. BellSoft Liberica will remain the default JVM.

Reviewer: My 2c here would be `BP_JVM_VENDOR` as this slots in with the existing `BP_JVM_TYPE` and `BP_JVM_VERSION`

> In addition to the current criteria for each buildpack, the detect method will also validate itself against the value of `BP_JAVA_JVM` and contribute only if it matches. BellSoft Liberica will remain the default JVM.

Author: I will change the name to `BP_JVM_VENDOR`.

> In addition to the current criteria for each buildpack, the detect method will also validate itself against the value of `BP_JVM_VENDOR` and contribute only if it matches. BellSoft Liberica will remain the default JVM.

Reviewer 2: ```suggestion
In addition to the current criteria for each buildpack, the detect method will also take the value of `BP_JVM_VENDOR` into account. If it is set, a buildpack should only detect if the value matches. BellSoft Liberica will remain the default JVM.
```

This would make it more clear, that a JVM providing buildpack should continue to detect, when all other criteria are being fulfilled, if there is no value provided for `BP_JVM_VENDOR`.
Although, this leads to the question of who will be providing `BP_JVM_VENDOR=liberica` to handle
> BellSoft Liberica will remain the default JVM.

Reviewer 3: The way we did this with `BP_JAVA_APP_SERVER` is that it's based on buildpack order. 

>* `$BP_JAVA_APP_SERVER` is `tomcat` or if `$BP_JAVA_APP_SERVER` is unset or empty and this is the first buildpack to provide a Java application server.

The way this works is that at build time, the buildpack checks to see if some other buildpack has already met that buildplan entry. If it has not been met, the buildpack is free to satisfy it.

Reviewer 2: I am not entirely sure I get this. The order is static, how are the alternative buildpacks meant to come into effect?

Anyway, your explanation does sound like it would still satisfy our requirement. As long as sap-machine keeps providing it's jvm dependencies even without `BP_JVM_VENDOR` set to `sap-machine`, we should be fine.

Reviewer 3: @loewenstein It works off the buildpack order defined in the composite buildpack.

So a proposed new order could be something like:

- bellsoft-liberica
- JVM vendor A
- JVM vendor B
- JVM vendor ...
- JVM vendor N
- <existing buildpack order starts here>

At detect, all of the buildpacks run. The JVM vendor providing buildpacks say "yes, I can provide you with a JVM". They all pass detection. The build process runs in sequence based on the order of buildpacks defined in the order group and that passed detection. Because all of the JVM vendor buildpacks pass by default without any user intervention, Bellsoft Liberica would run first and contribute the JVM. Then JVM Vendor A would run, it would see that the buildplan entry for a JDK/JRE has been satisfied and not do anything (just exit successfully). Then JVM Vendor B would do the same, and C and D, and so on. 

It's not a lot different than how things work now, if you just load multiple JVM vendor buildpacks into an order group. The main difference is that they would all ship this way by default, so a user wouldn't need to add `-b` flags to `pack build` and insert the JVM vendor buildpacks in the proper order.

Now, the other main difference is that a user can change this order by setting `BP_JVM_VENDOR=sap-machine` (or their vendor of choice. This works similarly. Detect runs in parallel, all the JVM vendor buildpack pass. The difference is at build time. The Bellsoft Liberica buildpack would run, it would see `BP_JVM_VENDOR=sap-machine` and opt-out, more specifically it'll mark the JRE/JDK buildplan entries as Unmet. This is a special condition that tells the lifecycle to essentially try another buildpack. The next JVM Vendor buildpack would run, look at `BP_JVM_VENDOR=sap-machine`, and so on until the SAP Machine buildpack runs. It would see `BP_JVM_VENDOR=sap-machine`, say that's me and install it's JRE/JDK. If there are any JVM vendor buildpacks left after it, they would then run and no-op (like in the default case).

So out of the box with this proposal, the paketo-buildpacks/java buildpack will ship with Bellsoft Liberica as the default, and it will include all of the other JVM vendors so a user can pick with `BP_JVM_VENDOR`. With the proposal as it stands, there is no way to change the default aside from changing the buildpack order which requires a custom composite buildpack (or repackage of paketo-buildpacks/java) or a custom builder that redefines the order groups (you don't need a composite buildpack, you could define the order groups at the builder level).

Having said that, there is an upstream buildpacks RFC that will allow for [builders to set environment variables](https://github.com/buildpacks/rfcs/pull/230). That has been approved but not implemented. When that is implemented, we would be able to consider adding some additional logic, perhaps a `BP_DEFAULT_JVM_VENDOR` (and similar for app servers) environment variable that would be intended to be set at the builder level. This could then further influence the logic, so rather than the default being the first, builder authors could more easily set a default. That's down the road in a different Paketo RFC though.

Reviewer 2: Alright, then it's about `build` not `detect`. But the same logic would still work, wouldn't it?
I.e. 
Liberica is first in the order, so while all jvm buildpacks detect, the build will skip early, because the build plan entry is already satisfied.
Looked at individually, the jvm buildpacks can all install as long as `BP_JVM_VENDOR` is either not set at all or set to their marker, e.g. `sap-machine`.
Our composite buildpack can keep packaging too different jvm buildpacks (sap-machine and a proprietary one) with non-overlapping versions and our users don't need to set `BP_JVM_VENDOR` at all.

Sound correct?

Author: That sounds correct and I will update the RFC to reflect the logic is in build, not detect.

Author: @loewenstein I have updated the RFC providing more details on how this is implemented.

| ------------       | ------------ 	| ------------------------------------|
| BellSoft Liberica  | liberica       |  bellsoft-liberica  buildpack       |
| Eclipse Openj9     | openj9         |  eclipse-openj9 buildpack           |
| Microsoft OpenJDK  | ms-openjdk     |  microsoft-openjdk buildpack        |
| Oracle             | oracle         |  oracle buildpack                   |
| Amazon Corretto    | corretto       |  amazon-corretto                    |
| Adoptium           | adoptium       |  adoptium buildpack                 |
| SAP Machine        | sapmachine     |  sap-machine buildpack              |
| Alibaba Dragonwell | dragonwell     |  alibaba-dragonwell buildpack       |
| GraalVM            | graalvm        |  graalvm buildpack                  |
| Azul Zulu          | azul-zulu      |  azul-zulu buildpack                |

Additional JVMs can easily be added to the Java buildpack by implementing this RFC.  

## Prior Art

- Offering alternate Java application servers in the composite java buildpack was implemented using an environment variable.  

## Unresolved Questions and Bikeshedding

> - None

Reviewer 3: So there is one technical hurdle here. Right now, the Paketo Builders (especially the full builder) are very close to the max layer limit for container images. This has temporarily put a hold on adding new buildpacks because each new buildpack pushes the layer limit up more.

There has been a proposal to squash down some of the layers, but I'm not sure where we are in terms of implementing that and how much of a reduction in the number of layers in the builders it would actually offer.

tl;dr - Adding ten more buildpacks isn't going to be feasible at this point.

The question I would propose is, do we wait for this to be resolved in the Paketo builders or do we strike out and make a Java-specific builder (this would require its own RFC also)? The Java-specific builder would have fewer layers as we'd remove the other language families.

Thoughts?

Author: Yes, I figured the technical hurdle would apply here but I at least wanted to get this discussion started. 

I do like the idea of a java-specific builder.  We don't have to add all the JVMs.  Do we know which are the most popular?

Reviewer 3: If we had a Java-specific builder, we could add all of the JVMs. That wouldn't be a problem.

## Motivation

There are a few motivations for this change:

1. Using an alternate JVM is tedious.  
2. The preferred JVM for the Liberty application server is eclipse openj9 because there are features in it that the liberty server takes advantage of we would like to exploit in the liberty buildpack.
3. The alternate JVM buildpack runs before the CA certs buildpack and therefore traffic from the alternate JVM vendor buildpack won’t trust any additional CA certs.  
4. Easier adoption of the other JVM buildpacks.

## Detailed Explanation

Currently, the Java buildpack only contributes the BellSoft Liberica JVM.  If we can accommodate multiple JVMs then we can provide a more
well-rounded java buildpack, reach more Java communities, generate more interest in Paketo and Cloud Native buildpacks, and ultimately make it easier for users to build images for Java applications.

## Rationale and Alternatives

Alternatives:

- Do nothing. Users of the Java buildpack will have to manually create a builder or manually specify the buildpack order, which is tedious.
- Create new composite buildpacks other JVMs as the default.

## Implementation

> In a composite buildpack all the JVM buildpacks will detect true to indicate they can provide a JRE/JDK.  At build time, 
> buildpacks execute in order, according to the order group.  The first JVM buildpack in the order group, bellsoft-liberica, will
> have the first opportunity to contribute the JVM regardless if `BP_JVM_VENDOR` is set.  If not set, the buildpack will be free
> to contribute.   If set, it should only contribute if `BP_JVM_VENDOR` is set to `liberica`.  Other buildpacks are given the 
> opportunity to provide the JVM and should only if the buildplan entry is not already met and the JVM name matches `BP_JVM_VENDOR`.  Otherwise, the buildpack will exit successfully without doing anything.  

Reviewer 2: ```suggestion
In a composite buildpack all the JVM buildpacks will detect true to indicate they can provide a JRE/JDK.  At build time, all JVM buildpacks are free to contribute if the buildplan entry is not met and `BP_JVM_VENDOR` is set to their tag or not set at all. Buildpacks execute in the order they are defined. The bellsoft-liberica buildpack comes first, which guarantees the current default behaviour.
```

I changed more than I intended, sorry. The main point is that for all buildpacks `BP_JVM_VENDOR` should either match their tag or not be set at all.

Author: @loewenstein Thanks for the suggestions.
